File: chroma-manager/chroma_ui_new/test/selenium/views/action-dropdown.js
'use strict';

var format = require('util').format;
var modalView = require('./modal');
var confirmActionModal = modalView.confirmActionModal;
var commandModal = modalView.commandModal;
var commandMonitor = require('../util/command-monitor');
var config = require('../config');
var msToSec = require('../util/time').msToSec;

/**
 * Gets an instance of an Actions Dropdown, tied to a specific server row.
 * @param {Object} dropdownContainer A server table row.
 * @returns {Object}
 */
function actionDropdownFactory (dropdownContainer) {

  var instance = {
    DROPDOWN: '.action-dropdown',
    get DROPDOWN_BUTTON () {
      return format('%s button', instance.DROPDOWN);
    },
    get DROPDOWN_MENU () {
      return format('%s .dropdown-menu', instance.DROPDOWN);
    },
    get dropdownButton () {
      return dropdownContainer.$(instance.DROPDOWN_BUTTON);
    },
    get dropdownMenu () {
      return dropdownContainer.$(instance.DROPDOWN_MENU);
    },
    /**
     * Click on action links a server row's Actions dropdown.
     * @param {String} actionName Name of the action (link) in the actions dropdown that we want
     * to invoke.
     */
    clickAction: function clickAction (actionName) {
      commandMonitor.waitForCommandsToFinish(config.wait_time.long);

      instance.waitForDropdownEnabled();

      instance.dropdownButton.click();

      instance.dropdownMenu.all(by.tagName('a'))
        .filter(function findItemsThatMatchAction (item) {
          return item.getInnerHtml()
            .then(function testForMatch (text) {
              return text === actionName;
            });
        })
        .then(function clickFiltered (filtered) {
          filtered[0].click();
        })
        .then(function checkForConfirm () {
          // If the action name was remove or shutdown, wait for it to complete.
          if ((actionName.indexOf('Remove') !== -1) || (actionName.indexOf('Shutdown') !== -1))
            confirmActionModal.waitForModal();

        })
        .then(function finish () {
          browser.isElementPresent(confirmActionModal.modal)
            .then(function isConfirmActionModalPresentThen (isPresent) {

              if (isPresent)
                confirmActionModal.confirmButton.click();

            })
            .then(function processCommandModal () {
              commandModal.waitForModal();
              commandModal.waitForCloseButtonToBeClickable();
              commandModal.closeButton.click();
              commandModal.waitForModalRemove(config.wait_time.medium);
            })
            .then(function processConfirmation () {

              browser.isElementPresent(confirmActionModal.modal)
                .then(function isConfirmActionModalPresentThen (isPresent) {

                  if (isPresent)
                    confirmActionModal.waitForModalRemove(config.wait_time.medium);

                });

            })
            .then(function waitForCommands () {
              commandMonitor.waitForCommandsToFinish(config.wait_time.long);
            });
        });

    },
    /**
     * Wait for this object's dropdown to be enabled.
     */
    waitForDropdownEnabled: function waitForDropdownEnabled () {

      browser.wait(
        instance.dropdownButton.isEnabled,
        config.wait_time.long,
        format('Timeout after %d seconds waiting for dropdown to be enabled', msToSec(config.wait_time.long))
      );

    }
  };

  return instance;
}

module.exports = actionDropdownFactory;

File: chroma-manager/tests/selenium/views/action_dropdown.py
from selenium.common.exceptions import StaleElementReferenceException, TimeoutException
from selenium.webdriver.support.wait import WebDriverWait
from tests.selenium.base_view import BaseView
from tests.selenium.utils.command_monitor import wait_for_commands_to_finish
from tests.selenium.views.modal import ConfirmActionModal, CommandModal


class ActionDropdown(BaseView):
    def __init__(self, driver, dropdown_container):
        super(ActionDropdown, self).__init__(driver)
        self.dropdown_container = dropdown_container

    DROPDOWN = '.action-dropdown'
    DROPDOWN_BUTTON = '%s button' % DROPDOWN
    DROPDOWN_MENU = '%s .dropdown-menu' % DROPDOWN

    @property
    def dropdown_button(self):
        return self.dropdown_container.find_element_by_css_selector(self.DROPDOWN_BUTTON)

    @property
    def dropdown_menu(self):
        return self.dropdown_container.find_element_by_css_selector(self.DROPDOWN_MENU)

    def find_action_by_text(self, text):
        return self.dropdown_menu.find_element_by_link_text(text)

    def wait_for_dropdown_enabled(self):
        return WebDriverWait(self, self.long_wait).until(
            lambda action_dropdown: action_dropdown.dropdown_button.is_enabled())

    def click_action(self, action_text):
        confirm_action_modal = ConfirmActionModal(self.driver)
        command_modal = CommandModal(self.driver)

        wait_for_commands_to_finish(self.driver, self.long_wait)

        self.dropdown_button.location_once_scrolled_into_view

        self.wait_for_dropdown_enabled()
        self.dropdown_button.click()
        action = self.find_action_by_text(action_text)
        action.click()

        modal = None

        try:
            modal = confirm_action_modal.wait_for_modal(self.short_wait)
        except TimeoutException:
            pass

        if modal:
            confirm_action_modal.confirm_button.click()

            command_modal.wait_for_modal()
            command_modal.wait_for_close_button_to_be_clickable()
            command_modal.close_button.click()
            command_modal.wait_for_modal_remove()

            confirm_action_modal.wait_for_modal_remove()

        wait_for_commands_to_finish(self.driver, self.long_wait)

        def wait_for_remove(action_dropdown):
            try:
                return action_dropdown.dropdown_button
            except StaleElementReferenceException:
                return False

        WebDriverWait(self, self.short_wait).until_not(wait_for_remove, "Action drop down was not removed.")
